cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java

```diff
@@ -0,0 +1,177 @@
+    this.maxRecordsToScanResolutionMap = maxRecordsToScanResolutionMap;
+    this.scheduledExecutorService = executorService;
+    this.cConf = cConf;
+  }
+
+  /**
+   * if the metrics data migration is already completed, we would just shutdown the executor service and return
+   * if not, we would schedule migration task at fixed rate schedule.
+   * when the migration is complete, the task would shutdown the executor service by itself
+   */
+  public void scheduleMigrationIfNecessary() {
+    if (!isMigrationComplete()) {
+      // we have 5 minute initial delay for availability of transaction/dataset service
+      // to be able to get the metrics table or if v3 tables are not already created they might take some time
+      // initially, we don't have to perform
+      scheduledExecutorService.scheduleAtFixedRate(new MigrationTask(), 5, 1, TimeUnit.MINUTES);
+      LOG.info("Scheduled metrics migration thread for resolution INT_MAX, 3600, 60 tables");
+    } else {
+      scheduledExecutorService.shutdown();
+    }
+  }
+
+  /**
+   * returns true when data migration is complete and
+   * all v2 metrics tables data has been migrated for configured resolutions ; false otherwise
+   * @return true if migration is complete; false otherwise
+   */
+  private boolean isMigrationComplete() {
+    Iterator<Integer> resolutions = maxRecordsToScanResolutionMap.keySet().iterator();
+
+    if (!resolutions.hasNext()) {
+      // all tables in migrationOrder map have been deleted, so migration is complete;
+      // useful when metrics.processor is running and completed migration
+      // and the run method removed all the entries from map when the tables were deleted
+      return true;
+    }
+
+    // when metrics.processor restarts and initializes DataMigrator,
+    // this is useful to decide whether to schedule migration thread or not
+    while (resolutions.hasNext()) {
+      int resolution = resolutions.next();
+      if (migrationTableUtility.v2MetricsTableExists(resolution)) {
+        try (MetricsTable v3MetricsTable = migrationTableUtility.getV3MetricsTable(resolution);
+             MetricsTable v2MetricsTable = migrationTableUtility.getV2MetricsTable(resolution)) {
+          if (v2MetricsTable == null || v3MetricsTable == null) {
+            // dataset/hbase service not available, skipping check this time, will try in next schedule.
+            // return false to try again in next attempt,
+            // by false we will proceed to run the migration, however if all tables have been migrated
+            // migrationTask can handle that scenario and shutdown the executor service.
+            return false;
+          }
+          MetricsTableMigration metricsTableMigration =
+            new MetricsTableMigration(v2MetricsTable, v3MetricsTable, cConf);
+          // if metrics data is available in old table, return false
+          if (metricsTableMigration.isOldMetricsDataAvailable()) {
+            return false;
```

Comment:
I am confused by this whole logic. Isn't this a lot easier: 
- schedule migration whenever any v2 table exists
- migration finds out whether there is data
  - if so, it migrates the data and then deletes the v2 table
  - if not, it deletes the table

I don't see much value in trying to determine here whether a table needs migration or deletion, and then have two different tasks for migration and deletion.